tservice/src/seal.rs

```diff
@@ -90,10 +90,74 @@ impl<T: AsRef<[u8]> + core::default::Default> SealedBuilder<T> {
     }
 }
 
+pub trait Unseal: AsRef<[u8]> {
+    /// The length (in bytes) needed to hold the decrypted text
+    fn decrypted_text_len(&self) -> Result<usize> {
+        let result = unsafe {
+            mc_sgx_tservice_sys::sgx_get_encrypt_txt_len(
+                self.as_ref().as_ptr() as *const sgx_sealed_data_t
+            )
+        };
+        match result {
+            // Per the documentation, UINT32_MAX indicates an error
+            u32::MAX => Err(Error::Unexpected),
+            size => Ok(size as usize),
+        }
+    }
+
+    /// Unseal the data in `self`
+    ///
+    /// Returns the unsealed data into the provided `buffer`.  The returned
+    /// slice will be the exact size of the unsealed data.
+    ///
+    /// # Arguments
+    /// * `buffer` - The buffer to output the unsealed data into.  `buffer`
+    ///   needs to be at least as big as [`decrypted_text_len`].
+    fn unseal<'a>(&self, buffer: &'a mut [u8]) -> Result<&'a mut [u8]> {
+        let data_length = self.decrypted_text_len()?;
+        if buffer.len() < data_length {
+            return Err(Error::InvalidParameter);
+        }
+
+        let mut data_length_u32 = data_length as u32;
+        let mut mac_length_u32 = 0;
+        unsafe {
+            mc_sgx_tservice_sys::sgx_unseal_data(
+                self.as_ref().as_ptr() as *const sgx_sealed_data_t,
+                ptr::null_mut(),
+                &mut mac_length_u32,
+                buffer.as_mut_ptr(),
+                &mut data_length_u32,
+            )
+        }
+        .into_result()?;
+
+        // While the lengths can be modified by `sgx_unseal_data`, we asked the
+        // SGX interface at the top of the function for the required sizes.
+        // If the sizes are different than we have unexpected behavior.
+        if data_length != data_length_u32 as usize || mac_length_u32 != 0 {
+            return Err(Error::Unexpected);
```

Comment:
It may be worth exercising this codepath in the tests.

Reply:
Unfortunately this logic we can't exercise without coming up with some kind of framework to run in the enclave.  
If one tries to compile a test for this crate it will fail to link up because it will try to pull in std for the test run, but the enclave code is no_std and when it links in the sgx libs it will be missing the allocator, panic handler, etc.

Reply:
Ah, that's irritating. In that case can you file an issue for future work about figuring out how to execute tests of this type? We're going to run into more of them in the future, I'm sure.

Reply:
Created, #135